macaw-dump/src/Data/Macaw/Dump/Discover.hs

```diff
@@ -0,0 +1,160 @@
+import What4.ProgramLoc qualified as WPL
+
+data DiscoverException = ELFResolutionError String
+  deriving Show
+
+instance X.Exception DiscoverException
+
+
+-- | Convert machine addresses into What4 positions.
+--
+-- When possible, we map to the structured 'WPL.BinaryPos' type. However, some
+-- 'MM.MemSegmentOff' cannot be mapped to an absolute position (e.g., some
+-- addresses from shared libraries are in non-trivial segments). In those cases,
+-- we map to the unstructured 'WPL.Others' with a sufficiently descriptive string.
+--
+-- TODO: import from `Testing`
+posFn :: (MM.MemWidth w) => Text.Text -> MM.MemSegmentOff w -> WPL.Position
+posFn binaryName segOff =
+  case MM.segoffAsAbsoluteAddr segOff of
+    Just mw -> WPL.BinaryPos binaryName (fromIntegral mw)
+    Nothing -> WPL.OtherPos (binaryName <> Text.pack ": " <> Text.pack (show segOff))
+
+-- | Load an ELF file into a macaw 'MM.Memory' (and symbols)
+--
+-- Prints warnings to stderr.
+loadELF ::
+  MML.LoadOptions ->
+  EE.ElfHeaderInfo w ->
+  IO (MM.Memory w, [MEL.MemSymbol w])
+loadELF loadOpts ehi = do
+  case MEL.resolveElfContents loadOpts ehi of
+    Left err -> X.throwIO (ELFResolutionError err)
+    Right (warnings, mem, _mentry, nameAddrList) -> do
+      F.forM_ warnings $ \w -> do
+        IO.hPutStrLn IO.stderr ("WARN: " ++ w)
+      return (mem, nameAddrList)
+
+-- | Run discovery on the provided symbols, or all if none are provided
+runDiscovery ::
+  forall arch w.
+  ( MC.ArchAddrWidth arch ~ w
+  , MC.ArchConstraints arch
+  , MM.MemWidth w
+  ) =>
+  EE.ElfHeaderInfo w ->
+  MAI.ArchitectureInfo arch ->
+  Set.Set BS.ByteString ->
+  IO (MD.DiscoveryState arch)
+runDiscovery headerInfo archInfo symbols = do
+  (mem, nameAddrList) <- loadELF MML.defaultLoadOptions headerInfo
+  let addrSymMap =
+        Map.fromList
+          [ (MEL.memSymbolStart msym, name)
+          | msym <- nameAddrList
+          , let name = MEL.memSymbolName msym
+          , Set.null symbols || Set.member name symbols
+          ]
+  pure (MD.cfgFromAddrs archInfo mem addrSymMap (Map.keys addrSymMap) [])
+
+displayCfgs ::
+  ( MC.ArchConstraints arch
+  , MS.GenArchInfo MS.LLVMMemory arch
+  , CCE.IsSyntaxExtension (MS.MacawExt arch)
+  ) =>
+  FilePath ->
+  MD.DiscoveryState arch ->
+  MS.GenArchVals mem arch ->
+  -- | Also print Crucible CFG?
+  Bool ->
+  IO ()
+displayCfgs path discState archVals printCrucible = do
+  let funInfos = discState Lens.^. MD.funInfo
+  halloc <- CFH.newHandleAllocator
+  F.for_ (Map.toList funInfos) $ \(_addr, Some info) -> do
+    IO.print (PP.pretty info)
+    Monad.when printCrucible $ do
+      let pos = posFn (Text.pack path)
+      let funName =
+            WF.functionNameFromText $
+              Text.decodeUtf8With Text.lenientDecode $
+                MD.discoveredFunName info
+      CCC.SomeCFG ssa <-
+        MS.mkFunCFG (MS.archFunctions archVals) halloc funName pos info
+      IO.print (CCC.ppCFG' True (CAP.postdomInfo ssa) ssa)
```

Comment:
Does it make sense to include some kind of headers to make it obvious which of the two `print`ed values correspond to which CFGs?